File: algorithms/138.copy-list-with-random-pointer.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        if (head == nullptr)  
            return nullptr;
            
        // 01. initialize the hash table & head node
        unordered_map<RandomListNode*, RandomListNode*> hash;
        RandomListNode *cur = head;
        hash[nullptr] = nullptr;
        hash[cur] = new RandomListNode(cur->label);
        
        // 02. add all nodes through next pointers
        while (cur && cur->next){
            hash[cur->next] = new RandomListNode(cur->next->label);
            hash[cur]->next = hash[cur->next];
            cur = cur->next;
        }
        
        // 03. copy the random table
        cur = head;
        while (cur){
            hash[cur]->random = hash[cur->random];
            cur = cur->next;
        }
        
        return hash[head];
    }
};

File: algorithms/138.copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil {
        return nil
    }

    result := map[*Node]*Node{nil: nil, head: &Node{Val : head.Val}}

    // 1. Copy the next pointers
    for cur := head; cur != nil  && cur.Next != nil; cur = cur.Next{
        result[cur.Next] = &Node{Val: cur.Next.Val}
        result[cur].Next = result[cur.Next]
    }

    // 2. Copy the random pointer
    for cur := head; cur != nil; cur = cur.Next{
        result[cur].Random = result[cur.Random]
    }
    
    return result[head]
}
